### PR TITLE
Feat/mpesa stk push

### PR DESCRIPTION
**Pull request by its-kios09**
## Requirements

- [ ] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [ ] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [ ] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

- (feat) added default value on the Bill Amount TextInput which allows the cashier either to initiate complete payment or partial depending on the patient's fund
- (feat) added a validator on the both phone number and bill amount
- (feat) added stk push for mobile money (Mpesa STK push)
- (feat) added a functionality to format phone number before initiating stk pusk can accept 07 or 01
- (feat) added a proposed for account number to be mflcode of facility#receipt to make the payment is for certain receipt for that will be paid off for security purpose to avoid the mpesa to be use on another facility example of the message is on the screenshot

## Screenshots

https://github.com/palladiumkenya/kenyaemr-esm-3.x/assets/67967749/1233e10a-622c-4eb0-90bd-f756959230cc

Screenshot_20240605-151914_Messages

*None.*

## Related Issue

*None.*

- Declaring of environment variables such as consumer key, passkey, consumer secret on .env file

## Other

*None.*

### CODE REVIEW

File: packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx [hunk 1]
```diff
@@ -1,11 +1,30 @@
+/* eslint-disable no-console */
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx:1]:
```suggestion

```

File: packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx [hunk 2]
```diff
@@ -1,11 +1,30 @@
...
+const InitiatePaymentSchema = z.object({
+  phoneNumber: z
+    .string({
+      required_error: 'Phone number is required',
+      invalid_type_error: 'Phone number must be numeric and 10 digits',
+    })
+    .refine((value) => /^\d{10}$/.test(value), {
+      message: 'Phone number must be 10 digits',
+    }),
+  billAmount: z.string({
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx:24]: I think bill amount will most likely be a number

File: packages/esm-billing-app/src/invoice/payments/payment.resource.ts [hunk 3]
```diff
@@ -1,6 +1,9 @@
+/* eslint-disable no-console */
 import useSWR from 'swr';
 import { openmrsFetch, useConfig } from '@openmrs/esm-framework';
 import { BillingConfig } from '../../config-schema';
+import { Buffer } from 'buffer';
+import axios from 'axios';
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/invoice/payments/payment.resource.ts:6]: Prefer to use native `fetch` API provided by the browser.

File: packages/esm-billing-app/src/invoice/payments/payment.resource.ts [hunk 4]
```diff
@@ -33,13 +36,34 @@ export const usePaymentModes = () => {
...
-  return openmrsFetch(url, {
-    method: 'POST',
-    body: payload,
-    headers: {
+export const generateStkAccessToken = async () => {
+  try {
+    const consumerKey = '';
+    const consumerSecret = '';
+    const authorizationUrl = 'https://sandbox.safaricom.co.ke/oauth/v1/generate?grant_type=client_credentials';
+    const auth = Buffer.from(`${consumerKey}:${consumerSecret}`).toString('base64');
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/invoice/payments/payment.resource.ts:44]:
```suggestion
    const auth = btoa(`${consumerKey}:${consumerSecret}`);
```

File: packages/esm-billing-app/src/invoice/payments/payment.resource.ts [hunk 5]
```diff
@@ -33,13 +36,34 @@ export const usePaymentModes = () => {
...
+    const headers = {
       'Content-Type': 'application/json',
-    },
-  });
+      Authorization: `Basic ${auth}`,
+    };
+    const response = await axios(authorizationUrl, { method: 'GET', headers });
+    const { access_token } = await response.data;
+    console.log(access_token);
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/invoice/payments/payment.resource.ts:51]:
```suggestion
  
```

File: packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx [hunk 6]
```diff
@@ -14,45 +33,91 @@ export interface InitiatePaymentDialogProps {
...
+          BusinessShortCode: shortCode,
+          Password: Password,
+          Timestamp: timeStamp,
+          TransactionType: 'CustomerPayBillOnline',
+          PartyA: phoneNumber,
+          PartyB: shortCode,
+          PhoneNumber: phoneNumber,
+          CallBackURL: callBackUrl,
+          AccountReference: accountReference,
+          TransactionDesc: 'HelloTest',
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx:89]:
```suggestion
          TransactionDesc: 'HelloTest',
```
I believe left by mistake

File: packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx [hunk 7]
```diff
@@ -14,45 +33,91 @@ export interface InitiatePaymentDialogProps {
...
+
+      validationErrors.forEach((error) => {
+        showSnackbar({
+          title: t('InitiatePaymentError', 'Initiate Payment Error'),
+          subtitle: error,
+          kind: 'error',
+          timeoutInMs: 3500,
+          isLowContrast: true,
+        });
+      });
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx:63]: No need to display snack bar for form validation, we should use form control invalidText for that. We might also want to disable the button to trigger STK push until we have a valid form state.

File: packages/esm-billing-app/src/invoice/payments/payment.resource.ts [hunk 8]
```diff
@@ -1,6 +1,9 @@
+/* eslint-disable no-console */
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/invoice/payments/payment.resource.ts:1]:
```suggestion

```

File: packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx (hunk 7 above)
Comment (ojwanganto) [reply packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx:63]: +1. The snackbar is a bit out of focus to a user.

**Comment by ojwanganto** [comment]
Thanks, @its-kios09. It is coming out well

File: packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx (hunk 2 above)
Comment (its-kios09) [reply packages/esm-billing-app/src/invoice/payments/initiate-payment/initiate-payment.component.tsx:24]: i will work on it

**Pull request closed without merging**